### PR TITLE
Agn features - DRW, EVT anomaly scores and PDRS 

### PR DESCRIPTION
**Pull request by zerozole**
# AGN / anomaly features for the SCoPe pipeline (`agn-features` branch)

## Motivation

SCoPe's feature engineering is tuned toward periodic and eclipsing variables.
This branch extends the feature set toward AGN and stochastic variability:
damped-random-walk (DRW) parameters, extreme-value-theory (EVT) anomaly
scores, and flare / high-activity segmentation. Light curves are thereby
marked for anomalies without touching any existing behavior: all new
features default to `include: false` (annotation-only), both new
feature-generation steps are behind opt-in flags, and no existing model,
feature file, or training set is affected.

Methods are adapted from two existing research codes:

- **PDRS** — (https://github.com/zerozole/Peak_Driven_Region_Segmentation) - peak-detection region segmentation of flares 
  (binned relative flux, gradient-guided region growing, saddle-point merging, region-median
  gate).
- **FLARE**  — (https://github.com/zerozole/FLARE) - DRW/OU baseline fitting
  with celerite and a peaks-over-threshold generalized Pareto calibration of
  the maximum standardized residual.

## What was added

| File | Change |
|---|---|
| `tools/featureGeneration/pdrs.py` | new: flare segmentation → `n_flares`, `flare_significance`, `flare_duty_cycle` |
| `tools/featureGeneration/drw.py` | new: celerite DRW/OU fit (RealTerm + JitterTerm, L-BFGS-B) → `drw_tau`, `drw_sigma`, `drw_max_z`, `drw_fit_success` |
| `tools/evt_calibration.py` | new: population-level GPD calibration → `anomaly_score`, `anomaly_flag` columns + frozen JSON calibration |
| `tools/generate_features.py` | `--doFlareDetection` / `--doDRW` flags (default off), computation hooked in after basic stats, hyperparameters read from config |
| `tools/generate_features_slurm.py` | forwards the two new flags |
| `config.defaults.yaml` | feature registration (`include: false`) + `feature_generation.pdrs/.drw/.evt` hyperparameter blocks |
| `requirements.txt`, `pyproject.toml` | `celerite>=0.4.2` dependency; `evt-calibration` console script |
| `tests/test_agn_features.py` | 22 synthetic-data tests (no Kowalski/GPU needed) |

`periodfind` is untouched.

## Method summary

**DRW features** (`--doDRW`): each light curve is fit with a celerite GP —
`RealTerm` (the DRW kernel; `sigma = sqrt(a)` is the asymptotic rms,
`tau = 1/c` the damping timescale) plus a `JitterTerm` absorbing
underestimated photometric errors — by maximizing the marginal likelihood
with L-BFGS-B. Standardized one-step-ahead residuals under the fitted OU
model give `drw_max_z`, the per-source anomaly statistic. A plain single
fit is the default (`max_iter: 1`); setting `max_iter > 1` enables FLARE's
iterative |z| > `z_thr` masking so flares don't contaminate their own
baseline. Note `SF_inf = sqrt(2) * drw_sigma` for comparison with the
quasar literature.

**PDRS features** (`--doFlareDetection`): magnitudes are converted to flux
relative to the median, binned (inverse-variance, 3-day bins), and searched
for peaks above `median + 2 sigma`; peaks are grown into regions with a
gradient-guided expansion, merged across shallow saddles, and gated on the
region median. Scalars: region count, max region significance, and the
fraction of the baseline spent in regions.

**EVT anomaly score** (`evt-calibration`, run after feature generation):
pools `drw_max_z` per filter, takes the 95th-percentile
peaks-over-threshold cut, fits a generalized Pareto distribution to the
exceedances (`floc=0`), and inverts it for a 1% target false-alarm rate.
`anomaly_score = -log10 P(max_z > z)`; `anomaly_flag = 1` above the
detection threshold. The calibration is frozen to JSON so later fields are
scored consistently. Dry-run by default; `--apply` writes the columns.

## Hyperparameters (defaults, tuned for ZTF DR23; all in config)

- `feature_generation.pdrs`: `bin_size_days: 3.0`, `peak_threshold: 2.0`,
  `smooth_window: 7`, `min_cluster_size: 3`, `saddle_ratio: 0.2`,
  `region_threshold: 0.5`, `max_gap_days: 60.0`
- `feature_generation.drw`: `min_epochs: 15`, `max_iter: 1`, `z_thr: 3.0`
- `feature_generation.evt`: `pot_percentile: 95.0`, `p_target: 0.01`

## Usage

```bash
generate-features --doFlareDetection --doDRW [usual arguments...]
evt-calibration --features-dir generated_features          # report only
evt-calibration --features-dir generated_features --apply  # write columns
```

## Testing

`python -m pytest tests/test_agn_features.py -v` — 22 tests, all passing:
DRW recovers known (tau, sigma) from exact OU simulations and its residuals
are ~N(0,1) under the true model; an injected flare raises `drw_max_z` from
~4 to ~9 and is flagged against a quiet DRW population after EVT
calibration; PDRS detects injected flares at significance > 10 versus < 4
on pure noise and keeps two well-separated flares as two regions; edge
cases (short curves, constant magnitudes, NaN handling, JSON round-trip)
are covered. The existing `tests/test_periodsearch.py` suite (87 tests)
still passes.

## Compatibility and later training use

Nothing needs retraining: all features are registered `include: false` and
both generation steps are opt-in. To use the features in classifiers later:
flip `include: true` in the config, regenerate features for the training
set with the flags on, rebuild the training set, and retrain (XGBoost
checks feature names and the DNN input width is fixed, so retraining is
mandatory at that point). Training, imputation, and inference all derive
feature lists from the config, so no further code changes are needed.

## Known caveats (candid notes for review)

- PDRS uses the light curve's *global* sigma both for the peak threshold
  and the region-median gate, so very active sources (several large flares)
  raise their own detection bar; in synthetic tests a multi-flare curve can
  gate out entirely while single flares are detected robustly. This is
  faithful to the original algorithm; if it matters for DR23,
  `region_threshold` is the knob to lower.
- With `max_iter: 1` (plain fit), flares slightly inflate their own DRW
  baseline, compressing the top of the `drw_max_z` scale; the EVT
  calibration is relative so ranking survives. `max_iter: 3` re-enables
  FLARE's masking at ~3x cost — worth an A/B check on a real field.
- AGN damping timescales near or above the survey baseline are poorly
  constrained (`drw_tau` saturates); `drw_max_z` / `anomaly_score` are the
  baseline-robust anomaly quantities.

### CODE REVIEW

**Comment by mcoughlin** [comment]
@zerozole can you give a sense of timing? Should this be in Rust / CUDA in periodfind?

**Comment by mcoughlin** [comment]
@zerozole please fix the linting

**Review by mcoughlin — APPROVED** [review]
This is great. LGTM
